reading_comprehension/qanet_for_drop.py

```diff
@@ -0,0 +1,330 @@
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_passage_vectors = util.weighted_sum(encoded_passage, attention_over_attention)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4)
+        merged_passage_attention_vectors = self._dropout(
+                torch.cat([encoded_passage, passage_question_vectors,
+                           encoded_passage * passage_question_vectors,
+                           encoded_passage * passage_passage_vectors],
+                          dim=-1)
+        )
+
+        modeled_passage_list = [self._modeling_proj_layer(merged_passage_attention_vectors)]
+
+        for _ in range(4):
+            modeled_passage = self._dropout(self._modeling_layer(modeled_passage_list[-1], passage_mask))
+            modeled_passage_list.append(modeled_passage)
+
+        passage_weights = self._passage_weights_predictor(modeled_passage_list[4]).squeeze(-1)
+        passage_weights = passage_weights * passage_mask
+        passage_vector = util.weighted_sum(modeled_passage_list[4], passage_weights)
+        question_weights = self._question_weights_predictor(encoded_question).squeeze(-1)
+        question_weights = question_weights * question_mask
+        question_vector = util.weighted_sum(encoded_question, question_weights)
+
+        # Shape: (batch_size, 3)
+        answer_type_logits = self._answer_type_predictor(torch.cat([passage_vector, question_vector], -1))
+        answer_type_probs = torch.nn.functional.softmax(answer_type_logits, -1)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 2 + modeling_dim))
+        passage_for_span_start = torch.cat([modeled_passage_list[1], modeled_passage_list[2]], dim=-1)
+        # Shape: (batch_size, passage_length)
+        span_start_logits = self._span_start_predictor(passage_for_span_start).squeeze(-1)
+        # Shape: (batch_size, passage_length)
+        span_start_probs = masked_softmax(span_start_logits, passage_mask)
+        # Shape: (batch_size, passage_length, encoding_dim * 2 + span_end_encoding_dim)
+        passage_for_span_end = torch.cat([modeled_passage_list[1], modeled_passage_list[3]], dim=-1)
+        # Shape: (batch_size, passage_length)
+        span_end_logits = self._span_end_predictor(passage_for_span_end).squeeze(-1)
+        # Shape: (batch_size, passage_length)
+        span_end_probs = masked_softmax(span_end_logits, passage_mask)
+
+        # Shape: (batch_size, # of numbers in the passage)
+        numbers_mask = util.get_text_field_mask(numbers_in_passage).float()
+        # Shape: (batch_size, # of numbers in the passage, embedding_dim)
+        embedded_numbers = self._dropout(self._text_field_embedder(numbers_in_passage))
+        embedded_numbers = self._number_embedding_proj_layer(embedded_numbers)
+        encoded_numbers = \
+            torch.cat([embedded_numbers, passage_vector.unsqueeze(1).repeat(1, embedded_numbers.size(1), 1)], -1)
```

Comment:
Instead of this concatenation, I was imagining something that does a linear prediction on the encoded representation of the passage index.  So, e.g.:

```python
encoded_numbers = modeled_passage_list[-1][number_indices]  # indexing isn't quite right here, but hopefully the idea makes sense
number_sign_logits = self._number_sign_predictor(encoded_numbers)
```

Does this make sense?  Seems easier for the model to just encode somewhere for each number whether it belongs in the computation, instead of having to remember that through a projection to a single passage vector.

Reply:
Yeah, this makes a lot of sense. I will do this.